src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -100,11 +103,35 @@ public async Task InitializedAsync()
         }
 
         [JsonRpcMethod(Methods.ShutdownName)]
-        public object? Shutdown(CancellationToken _) => null;
+        public Task ShutdownAsync(CancellationToken _)
+        {
+            Contract.ThrowIfTrue(_shuttingDown, "Shutdown has already been called.");
+
+            _shuttingDown = true;
+            _diagnosticService.DiagnosticsUpdated -= DiagnosticService_DiagnosticsUpdated;
+
+            return Task.CompletedTask;
+        }
 
         [JsonRpcMethod(Methods.ExitName)]
-        public void Exit()
+        public Task ExitAsync(CancellationToken _)
         {
+            Contract.ThrowIfFalse(_shuttingDown, "Shutdown has not been called yet.");
+
+            try
+            {
+                if (!_jsonRpc.IsDisposed)
+                {
+                    _jsonRpc.Dispose();
+                }
+            }
+            catch (Exception)
+            {
+                // Swallow exceptions thrown by disposing our JsonRpc object. Disconnected events can potentially throw their own exceptions so
+                // we purposefully ignore all of those exceptions in an effort to shutdown gracefully.
+            }
```

Comment:
Can we be more precise about the exceptions being caught here? I worry if we were catching exceptions that were our own fault and never knew.

Reply:
Not really to be honest. My goal here was to be over-protective to not crash VS. Not fully aware of the flavors of exceptions that can occur

Reply:
If we don't expect any exceptions here then we can report non-fatal exceptions to telemetry; that way we can catch if we're actually swallowing things that are actually bad.

Reply:
Can definitely do that. Would you mind linking an example of pushing these types of errors to your telemetry service?

Reply:
Just change the catch line to:

    catch (Exception e) when (FatalError.ReportWithoutCrash(e))